nixos/modules/services/databases/mysql.nix

```diff
@@ -438,72 +510,64 @@ in
-                      ${concatStringsSep "\n" (mapAttrsToList (database: permission: ''
-                        echo "GRANT ${permission} ON ${database} TO '${user.name}'@'localhost';"
-                      '') user.ensurePermissions)}
-                    ) | ${mysql}/bin/mysql -u root -N
-                  '') cfg.ensureUsers}
               '';
             in
               # ensureDatbases & ensureUsers depends on this script being run as root
               # when the user has secured their mysql install
               "+${setupScript}";
-          # User and group
-          User = cfg.user;
-          Group = "mysql";
-          # Runtime directory and mode
-          RuntimeDirectory = "mysqld";
-          RuntimeDirectoryMode = "0755";
-          # Access write directories
-          ReadWritePaths = [ cfg.dataDir ];
-          # Capabilities
-          CapabilityBoundingSet = "";
-          # Security
-          NoNewPrivileges = true;
-          # Sandboxing
-          ProtectSystem = "strict";
-          ProtectHome = true;
-          PrivateTmp = true;
-          PrivateDevices = true;
-          ProtectHostname = true;
-          ProtectKernelTunables = true;
-          ProtectKernelModules = true;
-          ProtectControlGroups = true;
-          RestrictAddressFamilies = [ "AF_UNIX" "AF_INET" "AF_INET6" ];
-          LockPersonality = true;
-          MemoryDenyWriteExecute = true;
-          RestrictRealtime = true;
-          RestrictSUIDSGID = true;
-          PrivateMounts = true;
-          # System Call Filtering
-          SystemCallArchitectures = "native";
-        };
-      };
+        }
+      ];
+    };
+
+    systemd.services.mysql-init = {
+      description = "MySQL database and user provisioning";
+
+      after = [ "mysql.service" ];
+      wantedBy = [ "mysql.target" ];
+      partOf = [ "mysql.target" ];
+
+      # ensureDatbases & ensureUsers depends on this script being run as root
+      # when the user has secured their mysql install
+      script = ''
+        ${optionalString (cfg.ensureDatabases != []) ''
+          (
+          ${concatMapStrings (database: ''
+            echo "CREATE DATABASE IF NOT EXISTS \`${database}\`;"
+          '') cfg.ensureDatabases}
+          ) | ${cfg.package}/bin/mysql -u root -N
+        ''}
+
+        ${concatMapStrings (user:
+          ''
+            ( echo "CREATE USER IF NOT EXISTS '${user.name}'@'localhost' IDENTIFIED WITH ${if isMariaDB then "unix_socket" else "auth_socket"};"
+              ${concatStringsSep "\n" (mapAttrsToList (database: permission: ''
+                echo "GRANT ${permission} ON ${database} TO '${user.name}'@'localhost';"
+              '') user.ensurePermissions)}
+            ) | ${cfg.package}/bin/mysql -u root -N
+          '') cfg.ensureUsers}
+      '';
+
+      serviceConfig = mkMerge [
+        commonServiceConfig
+        {
+          Type = "oneshot";
+          RemainAfterExit = true;
+        }
+      ];
```

Comment:
This service `mysql-init` need to run as from root?

Reply:
Yes : frown:

Reply:
@Izorkin currently `mysql57` and `mysql80` only provide `root` as an administrative user. I have a few ideas to "fix" this, though. I'll address that in a future PR because currently the module runs the `ExecStartPost` script as `root` anyways, so this PR isn't adding any additional `root` privileges.

Reply:
This variant also works:
```
diff --git a/nixos/modules/services/databases/mysql.nix b/nixos/modules/services/databases/mysql.nix
index e0573b1676a..8ef1377fdfa 100644
--- a/nixos/modules/services/databases/mysql.nix
+++ b/nixos/modules/services/databases/mysql.nix
@@ -17,6 +17,14 @@ let
   );

   commonServiceConfig = {
+    # User and group
+    User = cfg.user;
+    Group = cfg.group;
+    # Runtime directory and mode
+    RuntimeDirectory = "mysqld";
+    RuntimeDirectoryMode = "0755";
+    # Access write directories
+    ReadWritePaths = [ cfg.dataDir ];
     # Capabilities
     CapabilityBoundingSet = "";
     # Security
@@ -436,12 +444,6 @@ in
           Type = if isMariaDB then "notify" else "simple";
           Restart = "on-abort";
           RestartSec = "5s";
-          User = cfg.user;
-          Group = cfg.group;
-          RuntimeDirectory = "mysqld";
-          RuntimeDirectoryMode = "0755";
-          # Access write directories
-          ReadWritePaths = [ cfg.dataDir ];

           # The last two environment variables are used for starting Galera clusters
           ExecStart = "${cfg.package}/bin/mysqld --defaults-file=/etc/my.cnf ${mysqldOptions} $_WSREP_NEW_CLUSTER $_WSREP_START_POSITION";
```

Reply:
This would break users who have ran `mysql_secure_installation` because they would have no accounts they could authenticate against.

Reply:
Don't reproduced error with run `mysql_secure_installation`.
With mariadb `mysql_secure_installation`:
```
OTE: RUNNING ALL PARTS OF THIS SCRIPT IS RECOMMENDED FOR ALL MariaDB
      SERVERS IN PRODUCTION USE!  PLEASE READ EACH STEP CAREFULLY!

In order to log into MariaDB to secure it, we'll need the current
password for the root user. If you've just installed MariaDB, and
haven't set the root password yet, you should just press enter here.

Enter current password for root (enter for none):
ERROR 1698 (28000): Access denied for user 'root'@'localhost'
```
Working with `sudo mysql_secure_installation`.

With mysql working `mysql_secure_installation`:
```

Securing the MySQL server deployment.

Connecting to MySQL using a blank password.

VALIDATE PASSWORD PLUGIN can be used to test passwords
and improve security. It checks the strength of password
and allows the users to set only those passwords which are
secure enough. Would you like to setup VALIDATE PASSWORD plugin?

Press y|Y for Yes, any other key for No: y
```

Reply:
I have addressed this in https://github.com/NixOS/nixpkgs/pull/95231.